core/context-strategy/autoscale-context.ts

```diff
@@ -268,18 +288,65 @@ export class ConstantIntervalHeartbeatSyncStrategy implements HeartbeatSyncStrat
-            // delayAllowance: the time used in the calcualtion to offest any foreseeable latency
-            // outside of the function execution.
-            delay = heartbeatArriveTime - oldNextHeartbeatTime - delayAllowance;
+            delayAtArriveTime = heartbeatArriveTime - oldNextHeartbeatTime;
+            // if the device provide more information about the heartbeat, do a more accurate
+            // heartbeat calculation.
+            // check if the 'time' property exists. It can indicate the availability of the device
+            // info
+            const deviceSendTime: Date =
+                deviceSyncInfo.time !== null && new Date(deviceSyncInfo.time);
+            useDeviceSyncInfo = !!deviceSendTime;
+            if (useDeviceSyncInfo) {
+                delayCalculationMethod = 'by device send time';
+                // check if the sequence is in an incremental order compared to the data in the db
+                // if not, the heartbea should be marked as outdated and to be dropped
+                if (deviceSyncInfo.sequence < targetHealthCheckRecord.seq) {
+                    outdatedHearbeatRequest = true;
+                } else {
+                    // if the device seq is immediately after the recorded value
+                    // check if the send time match the heartbeat interval
+                    if (deviceSyncInfo.sequence === targetHealthCheckRecord.seq + 1) {
+                        const recordedSendTime: Date = new Date(targetHealthCheckRecord.sendTime);
+                        // compare using the device provided interval because if the interval
+                        // has changedthe new heartbeat will be sent in the new interval
+                        // this is the delay from the device's perspective
+                        delay =
+                            deviceSendTime.getTime() -
+                            recordedSendTime.getTime() -
+                            deviceSyncInfo.interval;
+                    }
+                    // NOTE:
+                    // in this situation, there are race conditions happening between some
+                    // heartbeat requests. The reason is one autoscale handler is taking much
+                    // longer to process another heartbeat and unable to complete before this
+                    // heartbeat arrives at the handler (by a parallel cloud function thread).
+                    // The outcome of this situation is:
+                    // for the hb which is immediate after the recorded one (new seq = old seq + 1)
+                    // it will be handled in the above if-else case.
+                    // for the other hb (new seq > old seq + 1), the delay cannot be calculated
+                    // thus discarding the delay calculation, and trust it is an on-time hb
+                    else {
+                        delay = -1; // on-time hb must have a negative delay
```

Comment:
seems like it would be cleaner to just have `null` and then handle the `null` case later. `-1ms` delay will be shown in the logs otherwise?

To properly account for this situation I guess you would need to store the last 3 heartbeat sequnce numbers in the database so you could retroactively count the missing sequence numbers?


10 -> 7 8 9 (+10): OK
12 -> 8 9 10 (+12): 1 missing...
11 -> 9 10 12 (+11): OK again

This is a simplified diagram for explanation. The final solution would need one record per heartbeat.. Could store the timestamp also so that could be (re)considered when the missing entries appear later.
(afterward delete them if the sequence number is lower than seq - 10 or so?)

This would allow you to calculate the loss count at each heartbeat based on multiple records from the db, rather than storing one record and overwriting each time and having issues with race conditions.

Reply:
This is a good idea! I believe since we now have the sequence from the sender, we can also make good use of the sequence. Let me think it again.